### PR TITLE
Refactor LSICorrelator as part of handover

### PR DESCRIPTION
**Pull request by JamesKingWork**
## PR Justification

Changes made as part of [#6871](https://github.com/ISISComputingGroup/IBEX/issues/6871) within this PR aim to take steps to refactor the LSI Correlator to decouple from PCASPY while allowing me to become more familiar with the LSI Correlator as part of the handover process which took place. 

## Work Completed
- Remove unused methods and imports.
- Place static variables inside config file to clean up scripts.
- Create a basic UML class diagram to describe existing structure of LSI correlator.
- Add more descriptive docstrings to better document codebase.
- Update README.md to include relevant and useful recourses.

## Acceptance Criteria
- [ ] All tests in `master\tests.py` pass.
- [ ] All IOC system tests pass in `EPICS-IOC_Test_Framework\tests\lsicorr.py`.
- [ ] Docstrings are well formatted and contain no spelling mistakes.
- [ ] README.md is well structured and grammatically correct.
- [ ] Static variables placed in config are well structured.
- [ ] Changes in `data_file_interaction.py` are clear and concise.

### How to Test
* Checkout the `Ticket6893_refactoring` branch of LSI Correlator Support repo in `C:\Instrument\Apps\EPICS\support\lsicorr\master` and pull recent changes down
* Checkout the `master` branch of the LSI Correlcator IOC repo in `C:\Instrument\Apps\EPICS\support\lsicorr_vendor\master` making sure most recent changes have been pulled down.
* run `master\tests.py` from the LSI Correlator Support repo and ensure all tests pass.
* Run IOC system tests in `EPICS-IOC_Test_Framework\tests\lsicorr.py`, ensuring all tests pass.
* Continue with the rest of the review for LSI Correlator Support repo

### CODE REVIEW

**Comment by JackEAllen** [comment]
@JamesKingWork could you add and commit the UML diagram you started in Draw.io and I'll modify it to work with the changes we've made on this so far?

**Comment by JamesKingWork** [comment]
> @JamesKingWork could you add and commit the UML diagram you started in Draw.io and I'll modify it to work with the changes we've made on this so far?

Added here https://github.com/ISISComputingGroup/EPICS-LSICorrelator/pull/11/commits/9664bdc539a61c85e0dc2b8115976d8890f8dd7a

File: correlator_pcaspy.py [hunk 1]
```diff
@@ -410,15 +355,14 @@ def get_user_filename(self):
...
-        ioc_name: The name of the IOC to be run, including ioc number (e.g. LSICORR_01)
-        pv_prefix: prefix for the pvs
-        macros: Dictionary containing IOC macros
-    Returns:
+    Serve the PVs for the remote ioc server forever.
 
+    @param ioc_name (str): The name of the IOC to run, including the ioc name (i.e. "LSICORR_01")
+    @param pv_prefix (str): The prefix to use for the PVs
+    @param macros (Dict[str, str]): The macros to use for the PVs - A dictionary containing IOC macros
+    @return: None
```
Comment (rerpha) [inline correlator_pcaspy.py:365]: i'd be tempted to remove the type hint in the comments in favour of the typing in the function signature so there's only one place you need to update it

File: correlator_pcaspy.py [hunk 2]
```diff
@@ -329,75 +282,67 @@ def take_data(self):
...
+            full_filename = os.path.join(Constants.DATA_DIR, filename)
         return full_filename
 
-    def get_user_filename(self):
+    def get_user_filename(self) -> str:
         """
         Returns a filename given the current run number and title.
 
         If device is simulated do not attempt to get run number or title from channel access
+        @return (str): Filename to save user data to
```
Comment (rerpha) [inline correlator_pcaspy.py:331]: here as well

File: correlator_pcaspy.py [hunk 3]
```diff
@@ -329,75 +282,67 @@ def take_data(self):
...
-
         # Set start PV back to NO, purely for aesthetics (this PV is actually always ready)
         self.update_param_and_fields(Records.START.name, 0)
         self.update_param_and_fields("{pv}:SP".format(pv=Records.START.name), 0)
 
     def get_metadata(self) -> Dict:
         """
-        Returns a dictionary containing meta data to be saved
+        Get the metadata to be saved with the data
+        @return (Dict): A dictionary containing meta data to be saved
```
Comment (rerpha) [inline correlator_pcaspy.py:297]: ""

File: correlator_pcaspy.py [hunk 4]
```diff
@@ -329,75 +282,67 @@ def take_data(self):
...
         return metadata
 
-    def get_archive_filename(self):
+    def get_archive_filename(self) -> str:
         """
         Returns a filename which the archive data file will be saved with.
 
         If device is simulated do not attempt to get instrument name or run number from channel access.
         If simulated save file in user directory instead of usual data directory.
+        @return (str): Filename to save archive data to
```
Comment (rerpha) [inline correlator_pcaspy.py:311]: ""

File: correlator_pcaspy.py [hunk 5]
```diff
@@ -232,62 +189,58 @@ def update_pv_and_write_to_device(self, reason: str, value: Any, update_setpoint
...
 
     @_error_handler
-    def read(self, reason: str):
+    def read(self, reason: str) -> Any:
         """
         Handle read of PV
-        Args:
-            reason (str): PV to read value of
+        @param reason (str): The name of the PV to get the value of (without the prefix)
+        @return (Any): The value of the PV
```
Comment (rerpha) [inline correlator_pcaspy.py:227]: ""

File: correlator_pcaspy.py [hunk 6]
```diff
@@ -232,62 +189,58 @@ def update_pv_and_write_to_device(self, reason: str, value: Any, update_setpoint
...
     @_error_handler
-    def write(self, reason: str, value: Any):
+    def write(self, reason: str, value: Any) -> None:
         """
         Handle write to PV
-        Args:
-            reason: PV to set value of
-            value: Value to set
+        @param reason (str): The name of the PV to set
+        @param value (Any): The new value to set the PV to
```
Comment (rerpha) [inline correlator_pcaspy.py:207]: ""

**Review by rerpha — CHANGES_REQUESTED** [review]
minor nitpicking, probably not really worth any change unless you really want to, just in future it may be best to just use type hinting in one way and avoid putting it in docstrings as it's more likely to go out of date if in both. 

Other than this i seem to get a unit test failure: 

```
FAIL: test_WHEN_save_data_THEN_metadata_written_AND_correlation_data_written_AND_traces_written (__main__.LSICorrelatorTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests.py", line 110, in test_WHEN_save_data_THEN_metadata_written_AND_correlation_data_written_AND_traces_written
    self.assertEqual(test_actual_data, file.read())
AssertionError: 'Pseudo Cross Correlation\nScattering angl[26064 chars]85\n' != '    Pseudo Cross Correlation\n    Scatter[26136 chars]85\n'
Diff is 26312 characters long. Set self.maxDiff to None to see it.
```
to me it looks like tabs are no longer stripped? 

ioc tests still pass which is good

**Comment by JackEAllen** [comment]
> minor nitpicking, probably not really worth any change unless you really want to, just in future it may be best to just use type hinting in one way and avoid putting it in docstrings as it's more likely to go out of date if in both.
> 
> Other than this i seem to get a unit test failure:
> 
> ```
> FAIL: test_WHEN_save_data_THEN_metadata_written_AND_correlation_data_written_AND_traces_written (__main__.LSICorrelatorTests)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "tests.py", line 110, in test_WHEN_save_data_THEN_metadata_written_AND_correlation_data_written_AND_traces_written
>     self.assertEqual(test_actual_data, file.read())
> AssertionError: 'Pseudo Cross Correlation\nScattering angl[26064 chars]85\n' != '    Pseudo Cross Correlation\n    Scatter[26136 chars]85\n'
> Diff is 26312 characters long. Set self.maxDiff to None to see it.
> ```
> 
> to me it looks like tabs are no longer stripped?
> 
> ioc tests still pass which is good

I've fixed the unit test now. Was a minor difference in the formatting of the Schema class when placed in the config file compared to the previous implementation.

File: correlator_pcaspy.py (hunk 1 above)
Comment (JackEAllen) [reply correlator_pcaspy.py:365]: Can I ask what you mean by replacing with function signatures? Could you provide a suggestion somewhere so I can see what you mean? Might move to using it if I think it looks nicer.
All methods include static types and return annotations. 
I decided to include the type hints just out of following a similar docstring format to that of other python projects I've worked on within ISIS for consistency and so the types are visible in intellisense next to the argument descriptions when previewing a method in a given IDE.

**Review by rerpha — APPROVED** [review]

**Pull request merged**